### PR TITLE
Add std nexttoward wrapper

### PR DESCRIPTION
**Pull request by yasahi-hpc**
This PR adds [nexttoward](https://en.cppreference.com/w/cpp/numeric/math/nextafter.html).
This function has following overloads

```C++
float nexttoward( float from, long double to );
double nexttoward( double from, long double to );
long double nexttoward( long double from, long double to );
template< class Integer >
double nexttoward( Integer from, long double to );
```

As the second argument is always `long double`, this is a host-only function.
In addition, some tests are skipped if `finite-math` is enabled

### CODE REVIEW

**Review by JBludau — APPROVED** [review]
No review, just to run SNL CI

File: core/unit_test/TestMathematicalFunctions.hpp [hunk 1]
```diff
@@ -3093,4 +3094,256 @@ TEST(TEST_CATEGORY, mathematical_functions_nextafter_fp16) {
...
+
+TEST(TEST_CATEGORY, mathematical_functions_nexttoward) {
+  if constexpr (std::is_same_v<TEST_EXECSPACE,
+                               Kokkos::DefaultHostExecutionSpace>) {
+    TestNextToward<int>();
+    TestNextToward<float>();
+    TestNextToward<double>();
+    TestNextToward<long double>();
+  } else {
+    GTEST_SKIP() << "This function has already been tested at host";
```
Comment (JBludau) [inline core/unit_test/TestMathematicalFunctions.hpp]:
```suggestion
    GTEST_SKIP() << "nexttoward only defined for host";
```

**Review by JBludau — APPROVED** [review]
I think this is fine. Since we are just forwarding to the std functions and this is host only the test seems pretty excessive. As soon as we add half type support we would have to add more testing though so I guess adding it here is fine.
I am still for splitting the test into floating point types and integer types (non blocking).

File: core/unit_test/TestMathematicalFunctions.hpp (hunk 1 above)
Comment (yasahi-hpc) [reply core/unit_test/TestMathematicalFunctions.hpp]: What about 

`GTEST_SKIP() << "Skip testing a host-only function nexttoward on device";`

**Review by cwpearson — COMMENTED** [review]
All the edge case tests look good. 

is it feasible to write one that has a hard-coded expected value and is not an edge case? I see you check that `nexttoward` and `nextafter` are in some way invertible, but I would like to see one test with a hard-coded value. Something like

```c++
    constexpr double from    = 3.7;
    constexpr long double to = 5.0L;
    constexpr double result   = std::nexttoward(from, to);
    // the last significand bit: ...999a → ...999b
    constexpr double expected = 0x1.d99999999999bp+1;

    static_assert(result == expected);
```
(godbolt is happy with this)

**Comment by JBludau** [comment]
> All the edge case tests look good.
> 
> is it feasible to write one that has a hard-coded expected value and is not an edge case? I see you check that `nexttoward` and `nextafter` are in some way invertible, but I would like to see one test with a hard-coded value. Something like
> 
> ```c++
>     constexpr double from    = 3.7;
>     constexpr long double to = 5.0L;
>     constexpr double result   = std::nexttoward(from, to);
>     // the last significand bit: ...999a → ...999b
>     constexpr double expected = 0x1.d99999999999bp+1;
> 
>     static_assert(result == expected);
> ```
> 
> (godbolt is happy with this)

Can you elaborate on what would you expect to gain from such a test in terms of information? We are just forwarding to functions in the std here (that on most cpus are mapped to intrinsics).  Wouldn't our tests either need to exhaust the  whole value range or we say that it is not on us to test the implementation in the std lib.

Also  https://en.cppreference.com/w/cpp/numeric/math/nextafter indicates these are only guaranteed to be `constexpr` with `c++23`

**Comment by yasahi-hpc** [comment]
> All the edge case tests look good.
> 
> is it feasible to write one that has a hard-coded expected value and is not an edge case? I see you check that `nexttoward` and `nextafter` are in some way invertible, but I would like to see one test with a hard-coded value. Something like
> 
> ```c++
>     constexpr double from    = 3.7;
>     constexpr long double to = 5.0L;
>     constexpr double result   = std::nexttoward(from, to);
>     // the last significand bit: ...999a → ...999b
>     constexpr double expected = 0x1.d99999999999bp+1;
> 
>     static_assert(result == expected);
> ```
> 
> (godbolt is happy with this)

I have roughly the same opinion as Jacob.
Concerning that current implementation is an alias to std, I think tests are good enough, which cover most cases

I think we need this kind of hard-coded tests for `half/bhalf` type, since `nexttoward` on `half/bhalf` type is our own implementation

**Comment by cwpearson** [comment]
Just a sanity check away from an edge case. I think the arguments you're making apply equally as well to the edge cases we're checking. I won't block on it either way.

**Review by cwpearson — APPROVED** [review]

**Comment by yasahi-hpc** [comment]
> Just a sanity check away from an edge case. I think the arguments you're making apply equally as well to the edge cases we're checking. I won't block on it either way.

That is true. May be too careful

File: core/src/Kokkos_MathematicalFunctions.hpp [hunk 2]
```diff
@@ -663,7 +663,32 @@ KOKKOS_IMPL_MATH_BINARY_INT_FUNCTION(scalbn, ldexp, int)
 KOKKOS_IMPL_MATH_UNARY_INT_FUNCTION(ilogb)
 KOKKOS_IMPL_MATH_UNARY_FUNCTION(logb)
 KOKKOS_IMPL_MATH_BINARY_FUNCTION(nextafter)
-// nexttoward
+inline float nexttoward(float from, long double to) {
```
Comment (nmm0) [inline core/src/Kokkos_MathematicalFunctions.hpp:666]: Should we keep the `constexpr`-ness with C++ >= 23?

Comment (dalg24) [reply core/src/Kokkos_MathematicalFunctions.hpp:666]: Good question but out of the scope of this PR.
At this point, we do not provide support for use in constant expression.

Comment (nmm0) [reply core/src/Kokkos_MathematicalFunctions.hpp:666]: Fine with me

**Review by nmm0 — APPROVED** [review]
